src/placement/jump_map.c

```diff
@@ -35,32 +35,114 @@ enum PL_OP_TYPE {
 	PL_ADD,
 };
 
-/**
- * Contains information related to object layout size.
- */
-struct jm_obj_placement {
-	unsigned int		jmop_grp_size;
-	unsigned int		jmop_grp_nr;
-	pool_comp_type_t	jmop_fdom_lvl;
-	uint32_t		jmop_dom_nr;
-};
+static void
+jm_obj_placement_fini(struct jm_obj_placement *jmop)
+{
+	if (jmop->jmop_pd_ptrs != NULL && jmop->jmop_pd_ptrs != jmop->jmop_pd_ptrs_inline)
+		D_FREE(jmop->jmop_pd_ptrs);
+}
 
-/**
- * jump_map Placement map structure used to place objects.
- * The map is returned as a struct pl_map and then converted back into a
- * pl_jump_map once passed from the caller into the object placement
- * functions.
- */
-struct pl_jump_map {
-	/** placement map interface */
-	struct pl_map		jmp_map;
-	/* Total size of domain type specified during map creation */
-	unsigned int		jmp_domain_nr;
-	/* # UPIN targets */
-	unsigned int		jmp_target_nr;
-	/* The dom that will contain no colocated shards */
-	pool_comp_type_t	jmp_redundant_dom;
-};
+/** Initialize the PDs for object to prepare for the layout calculation */
+#define LOCAL_PD_ARRAY_SIZE	(4)
+static int
+jm_obj_pd_init(struct pl_jump_map *jmap, struct daos_obj_md *md, struct pool_domain *root,
+	       struct jm_obj_placement *jmop)
+{
+	struct pool_domain	*pds, *pd;
+	uint8_t			*pd_used = NULL;
+	uint8_t			 pd_used_array[LOCAL_PD_ARRAY_SIZE] = {0};
+	uint32_t		 pd_array_size;
+	daos_obj_id_t		 oid;
+	uint64_t		 key;
+	uint32_t		 selected_pd, pd_id;
+	uint32_t		 pd_nr = jmap->jmp_pd_nr;
+	uint32_t		 dom_nr = jmap->jmp_domain_nr;
+	uint32_t		 shard_nr = jmop->jmop_grp_size * jmop->jmop_grp_nr;
+	uint32_t		 doms_per_pd;
+	uint32_t		 pd_grp_size, pd_grp_nr;
+	int			 i, rc;
+
+	jmop->jmop_root = root;
+	if (md->omd_pda == 0)
+		pd_nr = 0;
+	jmop->jmop_pd_nr = pd_nr;
+	if (jmop->jmop_pd_nr == 0) {
+		jmop->jmop_pd_grp_size = 0;
+		return 0;
+	}
+
```

Comment:
Probably we need a env setting for 2.4? only enable this if specific env are setting, to avoid it interfering sth for 2.4 release?

Reply:
now pool map only will add PD when with special fault_path config, so by default the PD code will not affect the placement generation.